File: apps/mobile/modules/local-access-privacy/android/src/main/java/expo/modules/localaccessprivacy/PrivacyVisibilityState.java
package expo.modules.localaccessprivacy;

// Java keeps this UI-independent state executable with the JDK, without an Android runtime.
// Authentication and background timing stay in the shared TypeScript service.
final class PrivacyVisibilityState {
  private boolean armed;
  private boolean foreground;
  private boolean visible;
  private boolean failed;
  private long generation;

  boolean isArmed() { return armed; }
  boolean isForeground() { return foreground; }
  boolean isFailed() { return failed; }
  long getGeneration() { return generation; }
  boolean isCovered() { return armed && (!foreground || !visible || failed); }
  boolean admitsForeground() { return foreground && (!armed || (!isCovered() && !failed)); }

  void arm() {
    armed = true;
    failed = false;
    cover();
  }

  void disarm() {
    armed = false;
    failed = false;
    cover();
  }

  void cover() {
    visible = false;
    generation += 1;
  }

  void setForeground(boolean value) {
    if (foreground != value) {
      foreground = value;
      cover();
    }
  }

  void fail() {
    failed = true;
    cover();
  }

  boolean publish(long expectedGeneration) {
    if (!armed || !foreground || failed || expectedGeneration != generation) {
      return false;
    }
    visible = true;
    return true;
  }

  boolean admitsAnnouncement(long expectedGeneration, boolean gate) {
    return expectedGeneration == generation
        && (!armed || (foreground && (gate || (!failed && visible))));
  }
}

File: apps/mobile/modules/local-access-privacy/ios/PrivacyVisibilityState.swift
// Native activity and visibility only. Authentication and background timing stay in TypeScript.
struct PrivacyVisibilityState {
  private(set) var armed = false
  private(set) var foreground = false
  private(set) var visible = false
  private(set) var failed = false
  private(set) var generation = 0

  var covered: Bool { armed && (!foreground || !visible || failed) }
  var admitsForeground: Bool { foreground && (!armed || (!covered && !failed)) }

  mutating func arm() {
    armed = true
    failed = false
    cover()
  }

  mutating func disarm() {
    armed = false
    failed = false
    cover()
  }

  mutating func cover() {
    visible = false
    generation += 1
  }

  mutating func setForeground(_ value: Bool) {
    guard foreground != value else { return }
    foreground = value
    cover()
  }

  mutating func fail() {
    failed = true
    cover()
  }

  mutating func publish(_ expectedGeneration: Int) -> Bool {
    guard armed, foreground, !failed, expectedGeneration == generation else { return false }
    visible = true
    return true
  }

  func admitsAnnouncement(_ expectedGeneration: Int, gate: Bool) -> Bool {
    expectedGeneration == generation && (!armed || (foreground && (gate || (!failed && visible))))
  }
}
